### PR TITLE
added puncher engage and disengage

### PR DESCRIPTION
**Pull request by NWSchallock**

### CODE REVIEW

File: src/Subsystems/Puncher.cpp [hunk 1]
```diff
@@ -32,6 +36,18 @@ void Puncher::SlingShot(double mSpeed)
 
 }
 
+double Puncher::GetPuncherLimit() {
+  return pullSling1.GetSensorCollection().GetPulseWidthPosition();
```
Comment (duane701) [inline src/Subsystems/Puncher.cpp]: `pullSling1.GetSensorCollection().GetPulseWidthPosition()` is not correct. Let's work out what this really is.

File: src/Subsystems/Puncher.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+{
+  MotorEngage(Puncher::kMotorEngage);
+  MotorDisengage(Puncher::kMotorDisengage);
+
+}
+
+void Puncher::InitDefaultCommand() {
+    // Set the default command for a subsystem here.
+    // SetDefaultCommand(new MySpecialCommand());
+  SetDefaultCommand(new::SlingShot(0.0));
```
Comment (duane701) [inline src/Subsystems/Puncher.cpp]: `new::SlingShot(0.0)` should be `new SlingShot(0.0)`

File: src/Subsystems/Puncher.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  return self;
+}
+
+Puncher::Puncher() : Subsystem(kSubsystemName),
+    pullSling1(RobotMap::kIDPullSling1),
+    pullSling2(RobotMap::kIDPullSling2),
+    releaseSling(RobotMap::kIDReleaseSling, RobotMap::kIDGetSling)
+{
+  MotorEngage(Puncher::kMotorEngage);
+  MotorDisengage(Puncher::kMotorDisengage);
```
Comment (duane701) [inline src/Subsystems/Puncher.cpp]: Remove `MotorDisengage(Puncher::kMotorDisengage)`;

File: src/Commands/SetSlingDisengagement.cpp [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#include "SetSlingDisengagement.h"
+
+SetSlingDisengagement::SetSlingDisengagement()
+{
+  Requires(Puncher::getInstance().get());
+}
+
+// Called once when the command executes
+void SetSlingDisengagement::Initialize() {
+  Puncher::getInstance()->MotorDisengage(Puncher::kMotorDisengage);
```
Comment (duane701) [inline src/Commands/SetSlingDisengagement.cpp]: Remove parameter `Puncher::kMotorDisengage`

File: src/Commands/SetSlingEngagement.cpp [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#include "SetSlingEngagement.h"
+
+SetSlingEngagement::SetSlingEngagement()
+{
+  Requires(Puncher::getInstance().get());
+}
+
+// Called just before this Command runs the first time
+void SetSlingEngagement::Initialize() {
+  Puncher::getInstance()->MotorEngage(Puncher::kMotorEngage);
```
Comment (duane701) [inline src/Commands/SetSlingEngagement.cpp]: Remove parameter `Puncher::kMotorEngage`

File: src/Subsystems/Puncher.cpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  }
+  return self;
+}
+
+Puncher::Puncher() : Subsystem(kSubsystemName),
+    pullSling1(RobotMap::kIDPullSling1),
+    pullSling2(RobotMap::kIDPullSling2),
+    releaseSling(RobotMap::kIDReleaseSling, RobotMap::kIDGetSling)
+{
+  MotorEngage(Puncher::kMotorEngage);
```
Comment (duane701) [inline src/Subsystems/Puncher.cpp]: Remove parameter `Puncher::kMotorEngage`

File: src/Subsystems/Puncher.h [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  enum EngageValue{kMotorEngage = frc::DoubleSolenoid::kForward,kMotorDisengage = frc::DoubleSolenoid::kReverse};
+
+  static std::shared_ptr<Puncher> getInstance();
+  Puncher();
+
+
+  void InitDefaultCommand();
+  void SlingShot(double speed);
+  double GetPuncherLimit();
+  void MotorDisengage(EngageValue value);
```
Comment (duane701) [inline src/Subsystems/Puncher.h]: Remove parameter `EngageValue value`

File: src/Subsystems/Puncher.h [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  void MotorEngage(EngageValue value);
```
Comment (duane701) [inline src/Subsystems/Puncher.h]: Remove `EngageValue value`

File: src/Subsystems/Puncher.cpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  pullSling1.Set(mSpeed);
+  pullSling2.Set(mSpeed);
+
+}
+
+double Puncher::GetPuncherLimit() {
+  return pullSling1.GetSensorCollection().GetPulseWidthPosition();
+}
+
+void Puncher::MotorDisengage(EngageValue value) {
```
Comment (duane701) [inline src/Subsystems/Puncher.cpp]: Remove parameter `EngageValue value`

File: src/Subsystems/Puncher.cpp [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+  releaseSling.Set(static_cast<DoubleSolenoid::Value>(value*-1));
```
Comment (duane701) [inline src/Subsystems/Puncher.cpp]: Should be

    releaseSling.Set(kMotorEngage)

File: src/Subsystems/Puncher.cpp [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+void Puncher::MotorEngage(EngageValue value) {
```
Comment (duane701) [inline src/Subsystems/Puncher.cpp]: Remove parameter `EngageValue value`

File: src/Subsystems/Puncher.cpp [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  releaseSling.Set(static_cast<DoubleSolenoid::Value>(value));
```
Comment (duane701) [inline src/Subsystems/Puncher.cpp]: Should be

    releaseSling.Set(kMotorDisengage)

**Review by duane701 — COMMENTED** [review]
FTP

**Comment by duane701** [comment]
Need to rebase in order to lose the conflicts.

We will do another PR to remove the second motor that we don't need.

**Comment by NWSchallock** [comment]
PTAL

**Review by duane701 — APPROVED** [review]
LGTM -- Rebase

I'll handle the rebase.

**Comment by duane701** [comment]
Replaced by #14

**Pull request closed without merging**